### DISCUSSION TITLE
Fetching all pages and data

### DISCUSSION THREAD

**Discussion by ricardobalk** [vuepress/core#72] [Q&A] [answered]
> TL;DR: How could someone fetch all pages and frontmatter data, like what was possible `this.$site.pages` in the previous version of VuePress?

---

I'm currently in the process of upgrading my personal website from VuePress to this newer VuePress 2, which uses Vue 3. At this point, I managed to [eject the theme in the new way](https://github.com/vuepress/vuepress-next/discussions/56) and [using the local theme with a relative path](https://github.com/vuepress/vuepress-next/issues/51). I also converted my Home page to VuePress 2, and it also loads the data provided via the frontmatter. So far, so good. :v:

The problem is... I'm stuck at the part of loading my blog posts, guides and portfolio items. Previously, I used `this.$site.pages` to [extract and filter all pages](https://github.com/ricardobalk/website/blob/a6e1f44517d038250abcb291d6e9cd09be4ea816/src/.vuepress/theme/mixins/retrievePosts.ts#L7-L23), as it contained a huge array with everything I needed.

(...) But what is the VuePress 2-equivalent of `this.$site.pages`?

---

Kind regards,

Ricardo

P.S. I already found that I could use the following line for the `Page` type:
```js
import type { Page } from '@vuepress/core/lib/types'
```

**Comment by yzx9** [comment, marked as answer]
TLDR: `import { usePagesData } from "@vuepress/client"`

I checked the following code to confirm that we are still provide `$site` on template:

https://github.com/vuepress/vuepress-next/blob/67118e8d9e57088e68076c631b1bfb58a6d0a730/packages/%40vuepress/client/src/app.ts#L129-L170

However, `$site.pages` no longer exists:

https://github.com/vuepress/vuepress-next/blob/67118e8d9e57088e68076c631b1bfb58a6d0a730/packages/%40vuepress/shared/src/types/site.ts#L7-L36

So, use hook usePagesData exported from client: 

```typescript
import { usePagesData } from "@vuepress/client"
```

Also a good idea to [inject pages global computed](https://github.com/nsznsznjsz/blog/blob/ef19b6cecef38cc10b6190b29a4d26162378b8f8/packages/theme-celesta/src/composables/usePosts.ts):

```typescript
// composable/usePosts.ts

import { inject } from "vue"
import { usePagesData } from "@vuepress/client"
import type { InjectionKey } from "vue"
import type { PageData } from "../types"

export const pagesSymbol: InjectionKey<Promise<PageData[]>> = Symbol("posts")

/**
 * Inject pages global computed
 */
export const usePages = async (): Promise<PageData[]> => {
  const pages= inject(pagesSymbol)
  if (!pages) {
    throw new Error("usePages() is called without provider.")
  }
  return pages
}

export const resolvePages = async (): Promise<PageData[]> => {
  const pagesData = usePagesData()

  const pages = await Promise.all(
    Object.keys(pagesData.value).map((key) => pagesData.value[key]())
  )

  return pages as PageData[]
}
```

```typescript
// clientAppSetup.ts
import { provide } from "vue"
import type { ClientAppSetup } from "@vuepress/client"
import { resolvePosts, postsSymbol } from "./composables"

const clientAppSetup: ClientAppSetup = () => {
  const pages = resolvePages()
  provide(pagesSymbol, pages)
}

export default clientAppSetup
```

**Reply by ricardobalk** [reply]
Alright, thanks for your quick response. 😄 

Regarding the 3rd line of `clientAppSetup.ts`, did you mean to use `import { resolvePages, pagesSymbol } from "./composables"`?

This is what I now have in my `filterPostsByType.ts`:
```ts
import type { PageData } from '@vuepress/core/lib/types'
import { reactive } from 'vue';
import { usePages } from '../composables';

/**
 * @returns A collection of posts filtered on the given type.
 * @param type Type of posts to return. Defaults to 'blogposts' when the type is an empty string, null or undefined
 * @example filterPostsByType('blogpost') - Returns all blog entries.
 * @example filterPostsByType('porfolio') - Returns all portfolio items.
 */
export function filterPostsByType(type: string): Array<PageData> {
  // If postType is null, undefined or an empty string,
  // use the default value of "blogpost".
  !type && (type = "blogpost");

  const results : PageData[] = reactive([])
  usePages().then((pages) => {
    pages.forEach(function(page) {
      page.frontmatter.type === type && results.push(page)
    })
  })
  return results
}

export default filterPostsByType
```

Seems that it's working so far...

![Screenshot from 2021-03-11 16-35-53](https://user-images.githubusercontent.com/14904229/110812541-fd7dad80-8287-11eb-9466-78697839e799.png)

<strike>The only thing that doesn't work yet is Vue's reactivity. I.e. it only shows the data in the template when editing the template. But that could also be caused by other errors. I could also remember something about pushing items into an array was a thing that Vue didn't handle well. Not sure 😅</strike> - Fixed, using `{ reactive } from 'vue'` 😃

**Reply by yzx9** [reply]
> Regarding the 3rd line of `clientAppSetup.ts`, did you mean to use `import { resolvePages, pagesSymbol } from "./composables"`?

Yes, this is a typo.

**Reply by yzx9** [reply]
Since `usePagesData()` is an async function, we need provide a promise. So, you can also return a promise:

```typescript
import type { PageData } from '@vuepress/core/lib/types'
import { reactive } from 'vue';
import { usePages } from '../composables';

/**
 * @returns A collection of posts filtered on the given type.
 * @param type Type of posts to return. Defaults to 'blogposts' when the type is an empty string, null or undefined
 * @example filterPostsByType('blogpost') - Returns all blog entries.
 * @example filterPostsByType('porfolio') - Returns all portfolio items.
 */
export async function filterPostsByType(type: string): Promise<Array<PageData>> {
  // If postType is null, undefined or an empty string,
  // use the default value of "blogpost".
  !type && (type = "blogpost");

  const pages = await usePages()
  return pages.filter(page => page.frontmatter.type === type)
}
```

Of course, you can return a reactive object, as you do now. However, this way will return an empty array at first time, and then there will be data, which may sometimes cause flash of unfilled content, so I tend to use async functions.

**Reply by ricardobalk** [reply]
@meteorlxy @nsznsznjsz - thank you both for your time, the thoughts and considerations on* the subjects. Awesome!

I am currently using the `filterPostsByType()` function inside my computed properties. By trial-and-error, I found out that Vue 3 does not support [Promise returns for computed properties](https://gist.github.com/loilo/fbe3124108a46ff50ab1b867bb5b4bf9), and instead wants a reactive object.

Citate from [Vue's official ESLint docs](https://eslint.vuejs.org/rules/no-async-in-computed-properties.html):

> Computed properties and functions should be synchronous. Asynchronous actions inside them may not work as expected and can lead to an unexpected behaviour, that's why you should avoid them.

Some sources suggest using [vue3-async-computed](https://www.npmjs.com/package/vue3-async-computed), but to me adding an extra dependency in order to make it work in an async way feels a bit unnecessary. Probably going to stick with my reactive object&mdash;based approach for now, but if one of you have an example snippet that works out of the box, without additional packages, that would be great.

P.S. I'm also working on making the function more universal, probably going to call it `filterPagesByFrontmatter()` and destructuring the parameters. I.e. converting it to something that allows me to call `filterPagesByFrontmatter({ type: "blogpost", lang: "en-US"})`

**Reply by yzx9** [reply]
@ricardobalk A little confusion, I did not fully understand your question, could you provide some code?

I saw you have an reactive object return, and filled it after promise fulfilled, so it should not be a asynchronous action, maybe like this:

```typescript
const posts = filterPagesByFrontmatter() // it is an empty array at first time
const data = computed(()=>{ return posts }) // this will fill by post after a timespan
```

Or you have using my version, you can await filter before computed properties: 

```typescript
const posts = await filterPagesByFrontmatter()
const data  = computed(() => { return posts })
```

Don't forgot they are all reactive!

**Reply by ricardobalk** [reply]
My apologies for the confusion. What I meant is that I made `filterPostsByType` return a promise, like you suggested, and was trying to use it in the following way, within `setup()` of my blog's main Vue component:
```js
const frontmatter = usePageFrontmatter<myThemePageFrontmatter>()
const posts = computed(async () => await filterPostsByType(frontmatter.value.type || "blogpost"))
```
If I remember correctly. But that doesn't work, obviously 🙈 

I also tried your example, as you just described, but as `setup()` isn't `async`, it gives the following error:
![image](https://user-images.githubusercontent.com/14904229/111792965-122ff600-88c5-11eb-9e56-236320dc3c81.png)

**Reply by yzx9** [reply]
@ricardobalk Sorry for answering your question so late because I temporarily left GitHub for a few days.

As your said, error is because `setup()` is not `async`, then the best way is also keep them `async`. Of course, you can take another way, such as set an default value to `posts`:

```typescript
const posts = ref([])
filterPostsByType(frontmatter.value.type || "blogpost")
  .then(filteredPosts => posts.value = filteredPosts)  // posts will be updated because it's reactive
return { posts }
```

**Reply by ricardobalk** [reply]
Cool! In the mean time, I have shared a separate branch for my transition to VuePress 2: https://github.com/ricardobalk/website/tree/feature/issues/%2387-move-to-vuepress-next. Most of it done, but still a lot of things to do. It might be better to open a separate thread for this.

With the [current version](https://github.com/ricardobalk/website/tree/6ec6fdcc10894eed7a25e61523f437d78358ef75) (as of writing), I get the following messages in the browser's console:

```
[Vue warn]: inject() can only be used inside setup() or functional components. runtime-core.esm-bundler.js:38
toRefs() expects a reactive object but received a plain one. reactivity.esm-bundler.js:783
[Vue warn]: inject() can only be used inside setup() or functional components. runtime-core.esm-bundler.js:38
[Vue warn]: Unhandled error during execution of setup function 
  at <Debug> 
  at <App> runtime-core.esm-bundler.js:38
Uncaught (in promise) Error: useSiteLocaleData() is called without provider.
    useSiteLocaleData siteLocaleData.js:6
    setup Debug.js:14
    callWithErrorHandling runtime-core.esm-bundler.js:154
    setupStatefulComponent runtime-core.esm-bundler.js:6537
    setupComponent runtime-core.esm-bundler.js:6498
    mountComponent runtime-core.esm-bundler.js:4201
    processComponent runtime-core.esm-bundler.js:4177
    patch runtime-core.esm-bundler.js:3788
    mountChildren runtime-core.esm-bundler.js:3970
    processFragment runtime-core.esm-bundler.js:4136
    patch runtime-core.esm-bundler.js:3781
    componentEffect runtime-core.esm-bundler.js:4293
    reactiveEffect reactivity.esm-bundler.js:42
    effect reactivity.esm-bundler.js:17
    setupRenderEffect runtime-core.esm-bundler.js:4258
    mountComponent runtime-core.esm-bundler.js:4217
    processComponent runtime-core.esm-bundler.js:4177
    patch runtime-core.esm-bundler.js:3788
    render runtime-core.esm-bundler.js:4878
    mount runtime-core.esm-bundler.js:3074
    mount runtime-dom.esm-bundler.js:1253
    <anonymous> client.js:16
siteLocaleData.js:6
[Vue warn]: Extraneous non-props attributes (class) were passed to component but could not be automatically inherited because component renders fragment or text root nodes. 
  at <Index class="theme-default-content custom" > 
  at <AsyncComponentWrapper class="theme-default-content custom" > 
  at <Content class="theme-default-content custom" > 
  at <About polaroidImage= 
Object { caption: "Photo of me in 2015", path: "/assets/img/content/home/ricardo.webp" }
 > 
  at <Home key=1 > 
  at <Layout> 
  at <AsyncComponentWrapper> 
  at <Vuepress onVnodeUnmounted=fn<onVnodeUnmounted> ref=Ref< 
<!--  -->
 > > 
  at <RouterView> 
  at <App> runtime-core.esm-bundler.js:38
[Vue warn]: Missing required prop: "links" 
  at <Footer> 
  at <Layout> 
  at <AsyncComponentWrapper> 
  at <Vuepress onVnodeUnmounted=fn<onVnodeUnmounted> ref=Ref< 
<!--  -->
 > > 
  at <RouterView> 
  at <App> runtime-core.esm-bundler.js:38
```

Hmm, hmm, this is a bit harder than I thought. Except for the "missing prop", I have no idea how to fix the `inject`, the `useSiteLocaleData() without provider` and the `Extraneous non-props attributes`. Where are they coming from?

Edit: `useSiteLocaleData()` concerns me the most, as it looks like it also blocking the `yarn build`

**Reply by yzx9** [reply]
Not sure if it is caused by an async setup function, maybe we need help @meteorlxy

**Comment by meteorlxy** [comment]
Since @nsznsznjsz has provided a possible solution, I would like to share the considerations behind this change.

Also provide a TLDR; version (lol): **Scalability**

---

Here's the original discussion:

- https://github.com/vuejs/vuepress/issues/2689#issuecomment-720602079
- https://github.com/vuejs/vuepress/issues/2689#issuecomment-720867853

In short, putting `pages` inside site data (i.e. `$site.pages`) will load all pages data when users enter your site, which is bad for large sites. (Imagine users may only visit 10 pages of your site, but they have to load all 1000 pages data)

The pages data is only useful for some features of VuePress v1:

- Built-in search: to build page indexes
- Blog plugin: to make a posts list

Those features are not required for all sites, so we make page data dynamically imported in v2 to try to avoid the scalability issue.

---

Now if we understand the problem, we might find that @nsznsznjsz 's [solution](https://github.com/vuepress/vuepress-next/discussions/72#discussioncomment-468833) should work well in small sites, but it would also encounter the same scalability issue as v1.

A better solution would be collecting data on demand in node side, and forwarding data to client side. For example, generate temp file in `onPrepared` hook and load it in client code:

```ts
// .vuepress/config.ts
export default {
  onPrepared: async (app) => {
    const myData = app.pages.map((page) => {
      // ...
    })
    await app.writeTemp('my-data.js', `export default ${JSON.stringify(myData)}`)
  },
}
```

```ts
// some-client-component.vue
import myData from '@temp/my-data'
```

However, this solution is more complicated, and should handle watching and HMR manually to get better dev experience. 🤔

**Reply by yzx9** [reply]
@meteorlxy Thanks for your detail reply and consideration such comprehensive! Scalability is indeed a crucial issue, `usePagesData()` should be an escape pod,  so we didn't write this API into docs on purpose?

**Reply by ricardobalk** [reply]
This might sound a bit strange, but would it be possible to provide this "fetch-data-from-all-pages" functionality in a separate package? So that people running VuePress for smaller blogs could use it as-is, and the caching functionality could be added later on during the process?

**Reply by meteorlxy** [reply]
@ricardobalk Of course. That should be useful for helping users migrating from v1.

Feature request & PR welcome.
